packages/myst-cli/src/project/fromTOC.ts

```diff
@@ -31,156 +32,44 @@ import { globSync } from 'glob';
-  return obj;
-}
+export function comparePaths(a: string, b: string): number {
+  const aDirName = dirname(a);
+  const bDirName = dirname(b);
+  // Same directory?
+  if (aDirName === bDirName) {
+    const aBaseName = basename(a);
+    const bBaseName = basename(b);
 
-/**
- * Take glob pattern and create nested pseudo-directory structure
- *
- * For example, given a pattern that resolves to:
- *
- * matches = [
- *   'one/two/three',
- *   'one/two/four',
- *   'five',
- * ]
- *
- * This builds:
- *
- * {
- *   one: {
- *     two: {
- *       three: {},
- *       four: {},
- *     },
- *   },
- *   five: {},
- * }
- */
-export function patternToDirectoryStructure(
-  pattern: string,
-  path: string,
-  opts?: Omit<Parameters<typeof globSync>[1], 'withFileTypes'>,
-) {
-  const matches = globSync(pattern, { cwd: path, ...opts });
-  const dirStructure: Record<string, any> = {};
-  matches.forEach((match) => {
-    // Glob always uses '/' separator, not filesystem separator
-    const fileParts = match.split('/');
-    objFromPathParts(fileParts, dirStructure);
-  });
-  return dirStructure;
-}
+    const aIsIndex = DEFAULT_INDEX_WITH_EXT.includes(aBaseName);
+    const bIsIndex = DEFAULT_INDEX_WITH_EXT.includes(bBaseName);
 
-/**
- * Take "directory structure" object and return corresponding toc entries
- *
- * The resulting entries are only FileEntries and ParentEntries. Similar
- * to the implicit project ordering, files come before parents.
- *
- * For example, given:
- *
- * dirStructure = {
- *   one: {
- *     two: {
- *       three: {},
- *       four: {},
- *     },
- *   },
- *   five: {},
- * }
- *
- * This builds:
- *
- * [
- *   { file: 'five' },
- *   {
- *     title: one,
- *     children: [
- *       {
- *          title: two,
- *          children: [
- *            { file: 'four' },
- *            { file: 'three' },
- *          ],
- *       },
- *     ],
- *   },
- * ]
- */
-export function directoryStructureToFileEntries(
-  dirStructure: Record<string, any>,
-  path: string,
-  ignore: string[],
-): (FileEntry | ParentEntry)[] {
-  let files = Object.keys(dirStructure)
-    // .map((file) => join(path, file))
-    .filter((file) => !ignore || !ignore.includes(join(path, file)))
-    .sort(sortByNumber);
-  let indexFile: string | undefined;
-  DEFAULT_INDEX_WITH_EXT.forEach((index) => {
-    if (indexFile) return;
-    indexFile = files.find((file) => file.toLowerCase() === index);
-  });
-  if (indexFile) {
-    files = [indexFile, ...files.filter((file) => file !== indexFile)];
+    // Are both a and b index files, or neither?
+    if (aIsIndex === bIsIndex) {
+      return aBaseName.localeCompare(bBaseName, undefined, {
+        numeric: true,
+        ignorePunctuation: true,
+      });
+    }
+    // Prefer index files
+    else if (aIsIndex) {
+      return -1;
+    } else {
+      return +1;
+    }
+  }
+  // Otherwise, compare naively as strings
+  else {
+    return a.localeCompare(b, undefined, {
```

Comment:
Another difference from implicit toc behavior: Here, folders and files get mixed together alphabetically; in the implicit toc, in a given directory, folders are always after files.

I feel like this one is totally fine. Worse case scenario, folders can always be renamed to order them differently.

Reply:
We can make this a choice. I'm perfectly happy with either solution. I perhaps slightly prefer the current logic, as most file-system browsers e.g. `ls` use this kind of ordering.